src/Shared/JSInterop/JSCallResultTypeHelper.cs

```diff
@@ -1,15 +1,39 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
+using System.Reflection;
+
 namespace Microsoft.JSInterop
 {
     internal static class JSCallResultTypeHelper
     {
+        // We avoid using Assembly.GetExecutingAssembly() because this is shared code.
+        private static readonly Assembly _currentAssembly = typeof(JSCallResultType).Assembly;
+
         public static JSCallResultType FromGeneric<TResult>()
-            => typeof(TResult) == typeof(IJSObjectReference)
-            || typeof(TResult) == typeof(IJSInProcessObjectReference)
-            || typeof(TResult) == typeof(IJSUnmarshalledObjectReference) ?
-                JSCallResultType.JSObjectReference :
-                JSCallResultType.Default;
+        {
+            var resultType = typeof(TResult);
+
+            if (resultType.Assembly == _currentAssembly)
+            {
+                if (resultType == typeof(IJSObjectReference)
+                    || resultType == typeof(IJSInProcessObjectReference)
+                    || resultType == typeof(IJSUnmarshalledObjectReference))
+                {
+                    return JSCallResultType.JSObjectReference;
+                }
+                else
+                {
+                    throw new ArgumentException(
```

Comment:
Wouldn't this trip on `DotNetObjectReference`? Should we check if the type derives from `IJSObjectReference`?

Reply:
Good catch. Looks like the current tests didn't catch this because they wrap `DotNetObjectReference` in a struct, so the generic parameter bypasses these checks.

Another thing I've realized is that if the customer creates their own class deriving from `IJSObjectReference` and attempts to grab an instance of it via JS interop, the new exception will not be thrown since their own type is from a different assembly. Should we handle this case as well?

It seems that for the exception to be thrown if and only if the user is asking for an `IJSObjectReference` type that is not one of the three allowed types, we would have to:
1. Check if the generic parameter type can be assigned to `IJSObjectReference` (something like `typeof(IJSObjectReference).IsAssignableFrom(resultType)`).
2. Check if the generic parameter type does not match one of the three accepted types.

But at that point, I'm not sure the new solution is really better than we we already had (@SteveSandersonMS expressed performance concerns with `Type.IsAssignableFrom` in a previous PR).

Thoughts?

Reply:
> Another thing I've realized is that if the customer creates their own class deriving from IJSObjectReference and attempts to grab an instance of it via JS interop, the new exception will not be thrown since their own type is from a different assembly. Should we handle this case as well?

I think that should be fine. If they're dedicated enough to subtype this, we'll let them manage it's use. Could we simply check for the presence of the types we know users could accidentally misuse - `JSObjectReference` , and `JSInProcessObjectReference` and throw?

Reply:
> Another thing I've realized is that if the customer creates their own class deriving from IJSObjectReference and attempts to grab an instance of it via JS interop, the new exception will not be thrown since their own type is from a different assembly. Should we handle this case as well?

That doesn't seem like a problem to me either. We don't need to recognize 3rd-party subtypes of `IJSObjectReference` since there's no way we'd know which concrete instance type to create. So I'd expect it to be treated as `JSCallResultType.Default`, as it already is here.

Good catch on `DotNetObjectReference` though! To be honest, I'm not massively bothered about providing a specific exception if people try to use `JSObjectReference` instead of `IJSObjectReference`. The equivalent would be trying to `@inject JSRuntime JS` instead of `IJSRuntime`, and we don't have any exception for that except the default one from DI.

I think if someone does mistakenly use `InvokeAsync<JSObjectReference>(...)` then they'll get an exception from the JSON deserializer saying it doesn't have a default constructor, which is basically good enough because then the developer will have no choice but to realise they are doing it wrong and will have to check docs again or another working implementation to copy/paste.

So in summary I'd be fine with a simplified update like:

```cs
if ((resultType.Assembly == _currentAssembly)
  && (resultType is one of the three known IJSObjectReference ones))
{
    return JSCallResultType.JSObjectReference;
}
else
{
    return JSCallResultType.Default;
}
```

We could even choose to leave the previous implementation as-is, but adding the `resultType.Assembly == _currentAssembly` check seems at least marginally useful.